doc/design/shattered_gpu_resource_detector/shattered_gpu_resource_detector.md

```diff
@@ -0,0 +1,78 @@
+# 碎片化资源估算框架设计
+
+## 现状分析
+
+目前我们在使用集群的时候将集群看做是一个大机器来进行资源分配，但实际情况是，集群和大机器还是会有本质上的区别，这个区别在于，如果没有合理的调度 & 再调度能力的支撑，集群上的机器很容易会生成碎片资源，从而导致在外界看来，集群还有相当部分的资源，然而对应的大任务无法被调度到集群上。
```

Comment:
I think this is something EDL is doing right now from my understanding. can you give more specific and in detail description of this proposal please?

Reply:
Currently we are suffering from the shattered gpu resource in our online paddle training cluster, so we need to figure out a way to detect and process it by using the rescheduler with the help of edl's fault-tolerant and auto-scale capability. We recently refactored edl to support Kubernetes crd, and I think it should be the good time to start resolving the issue described in the pull request.

Reply:
I see, so this is mainly focusing on a finer granularity of the computing resource evaluation, especially on the GPU resource? do I understand this correctly?
if so,  I have 2 follow-up questions: 
1. can you give an example of "shattered gpu resource"?
1. at the end of the day, the scheduling is handled by Kubernetes itself, how would this evaluator optimize the scheduling of kubernetes?
1. if the job can not be deployed to a cluster, it's marked with `PENDING` and wait in queue. is this some kind of equivalence of job evaluation?

Reply:
1. This is the real scenario in current Paddle online/offline environment in Baidu: the user have 8 gpu card quota remaining in his/her account, so he/she submit a 8-card paddle training job, but in the cluster, those 8 cards located on 4 different machines, with each machine have 2 remaining card, this is what we called the shattered gpu resource.

2. I submitted another pull request about the basic idea of rescheduler, which will cooperate with the implementation of this pull request, the process will be: the detector detects that there are shattered gpu resource in the cluster, that prevent the new training job from being scheduled, then it will hold this training job in a high priority queue, and triggered the rescheduler to optimize the gpu tasks in the cluster by leveraging on the fault-tolerant and scale capability in edl. After the reschedule, the cluster might have enough space for the training job to get allocated.

3. We wrote our own paddle job controller, and yes, the pod will be marked as pending in the cluster, and will have k8s event like "no sufficient gpu resource for job xxx", this will confuse the end user, he/she have enough quota for the training job, but the system told him/her that there is no sufficient gpu resource, which we think is not good enough in our current design.

Reply:
I got it, this is a valid case to improve EDL, thanks for the great idea. Can you add the examples you mentioned above to this doc so that other people can understand this scenario as well? 
Also, since we have the "training job" defined as CRD, we can further customize the definition of this CDR to have finer granularity on the resource requirement part, for example, when a job says it needs 8 GPUs, and we have 4 nodes with 2 available GPUs each, in this case we can start 4 pods as trainers to fully utilize the shattered GPU resource. my point is: our rescheduling logic should have 2 stages, 1. try to optimize new job first, 2. then try to reschedule the existing jobs in the cluster to make room for the new job.

Reply:
```
Can you add the examples you mentioned above to this doc so that other people can understand this scenario as well?
```
Sure, I'll add additional examples for people to understand it.

```
my point is: our rescheduling logic should have 2 stages, 1. try to optimize new job first, 2. then try to reschedule the existing jobs in the cluster to make room for the new job.
```
That makes sense to me, I'll modify the doc and hopefully you can review it for me.

Reply:
sure, thanks